core/src/test/scala/unit/kafka/log/LogLoaderTest.scala

```diff
@@ -352,19 +352,15 @@ class LogLoaderTest {
       // Intercept all segment read calls
       val interceptedLogSegments = new LogSegments(topicPartition) {
         override def add(segment: LogSegment): LogSegment = {
-          val wrapper = new LogSegment(segment) {
-
-            override def read(startOffset: Long, maxSize: Int, maxPosition: Long, minOneMessage: Boolean): FetchDataInfo = {
-              segmentsWithReads += this
-              super.read(startOffset, maxSize, maxPosition, minOneMessage)
-            }
-
-            override def recover(producerStateManager: ProducerStateManager,
-                                 leaderEpochCache: Optional[LeaderEpochFileCache]): Int = {
-              recoveredSegments += this
-              super.recover(producerStateManager, leaderEpochCache)
-            }
-          }
+          val wrapper = Mockito.spy(segment)
```

Comment:
Using a `spy` mock instead of a copy constructor is not better. What is the purpose of this change? cc @chia7712

Reply:
> What is the purpose of this change?

It seems to me there are two reason of removing the copy constructor.

1. it does not copy all fields of `LogLoader`, so it is potential bug if we use it in the production (in the future)
2. the copy constructor is used only in testing, so it should be fine to move the related behavior to testing code

> Using a spy mock instead of a copy constructor is not better. 

pardon me, could you please share the reason to me?

Reply:
There is a lot of magic to make mocks work and even more so for spy mocks. Replacing readable/debuggable normal code for mocks is generally a bad idea. A concrete example or the problems with mocks are the mock leaks that have caused serious test instability.

As a general rule, mocks should only be used if the mock free alternative is clearly worse in terms of verbosity, maintainability, etc. I don't think it's the case here.

Reply:
> There is a lot of magic to make mocks work and even more so for spy mocks. Replacing readable/debuggable normal code for mocks is generally a bad idea. A concrete example or the problems with mocks are the mock leaks that have caused serious test instability.
As a general rule, mocks should only be used if the mock free alternative is clearly worse in terms of verbosity, maintainability, etc. I don't think it's the case here.

Thanks for sharing. I agree that `spy` is a black box and it could have caused serious test instability.

get back to the subject. There are alternatives which can remove the incomplete copy constructor.

1. add public getters to `LogSegment` in order to reuse the constructor (https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java#L112). The side-effect is that the testing-only code is added. However, that is better than having an incomplete copy constructor.
2. reuse the constructor with temporary objects. The is the initial approach of this PR. The disadvantage is that it needs to create weird objects, but it is fine for the tests.

I prefer option 2 to replace the `spy`. WDYT?